### PR TITLE
Issue549repair

### PR DESCRIPTION
**Pull request by plwhite**
Fixes issue #549, but fixes IP v6 and adds many tests.

### CODE REVIEW

**Comment by plwhite** [comment]
Note that there are no changes to devices since the last iteration, but I tidied up the code in endpoints fixing the breaking of IP v6, and wrote enough tests to cover every line that I had changed.

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # Changelog
 
+
 # 0.23-dev
+- Reset ARP configuration when endpoint MAC changes.
+- Firewall input characters in profiles and tags.
```
Comment (matthewdupre) [inline CHANGES.md]: This line is already in the changelog for 0.22.

File: calico/felix/endpoint.py [hunk 2]
```diff
@@ -339,37 +362,43 @@ def _remove_chains(self):
             _log.exception("Failed to delete chains for %s", self)
             self._failed = True
 
-    def _configure_interface(self):
+    def _configure_interface(self, mac_changed):
         """
         Applies sysctls and routes to the interface.
+
+        :param: bool mac_changed: Has the MAC address changed since it was last
+        configured? If so, we reconfigure ARP for the interface.
```
Comment (matthewdupre) [inline calico/felix/endpoint.py]: This comment should specify IPv4 (I'd quite like to see an explanation of why we can't reset ARP for IPv6 too).

File: calico/felix/test/test_endpoint.py [hunk 3]
```diff
@@ -32,21 +33,153 @@
...
+                result = local_ep.on_endpoint_update(data, async=False)
+                self.assertEqual(local_ep._mac, data['mac'])
+                devices.configure_interface_ipv4.assert_called_once_with(iface)
+                devices.set_routes.assert_called_once_with(ip_type,
+                                                           set(ips),
+                                                           iface,
+                                                           data['mac'],
+                                                           reset_arp=False)
+
+        data['mac'] = stub_utils.get_mac()
```
Comment (matthewdupre) [inline calico/felix/test/test_endpoint.py]: These blocks are pretty scary looking, and it's not clear what the test is doing unless you know exactly what the code   it's testing does.  Please add comments (short ones are fine) to say what each step is testing.

File: CHANGES.md (hunk 1 above)
Comment (plwhite) [inline CHANGES.md]: Ouch. Merge error on my part.

**Comment by matthewdupre** [comment]
@plwhite - guess what - it's still broken; http://calico-jenkins:8080/job/run-fv-tests/901/console

I've reverted the merge again.  Since this has gone wrong twice already, please can you manually run the FVs over this before resubmitting it.

Thanks

**Pull request merged**